### PR TITLE
Fix leak and BOM handling issues in DxcCreateBlob

### PR DESCRIPTION
**Pull request by tex3d**
A leak would happen when input buffer is considered an emptyString: zero-length or a string with only a BOM and/or null-terminator. Fixed by freeing the supplied buffer in the emptyString case when we are supposed to take ownership of it.

BOM handling was also problematic because it offset the original pPtr input to the function. If we were taking ownership of this buffer, this was not safe, since the pointer could no longer be used with Free.

Fix is:
- Keep track of pContents and contentSize separately, adjusting those according to the BOM.
- Use pContents and contentSize for bPinned and bCopy cases.
- Wrap blob in blob to skip BOM when taking ownership of memory.

Also:
- moved null-termination into copy block, instead of next to blob creation.

### CODE REVIEW

File: lib/DxcSupport/FileIOHelper.cpp [hunk 1]
```diff
@@ -668,34 +678,50 @@ HRESULT DxcCreateBlob(
...
-    else
-      memset(pData, 0, size);
+    DXASSERT_NOMSG(pContents && contentSize);
+    memcpy(pData, pContents, contentSize);
+
+    // Guarantee null-termination
+    if (bNullTerminated && newSize > contentSize) {
+      if (codePage == CP_UTF8)
+        ((char*)pData)[newSize - 1] = 0;
+      else if (codePage == DXC_CP_WIDE)
```
Comment (dmpots) [inline lib/DxcSupport/FileIOHelper.cpp:688]: What if codePage is neither CP_UTF8 or DXC_CP_WIDE? Is that impossible? Could add an else { assert(false) } in that case.

**Review by dmpots — APPROVED** [review]
The leak case will not be caught by the current test infrastructure (requires leak tracking).

But the BOM case could have been (we should have crashed freeing a pointer that was not returned from Alloc()). We should add some test coverage for the BOM case and probably good to flesh out the empty string and zero-length tests (if needed) as well.

**Comment by damyanp** [comment]
This PR was closed as it has not been updated in the last two years. Please feel free to reopen if this PR should be merged and is in a reviewable state.

**Comment by damyanp** [comment]
This PR was closed as it has not been updated in the last two years. Please feel free to reopen if this PR should be merged and is in a reviewable state.
